### PR TITLE
deduplicate dns record in lookup

### PR DESCRIPTION
**Pull request by lyuxuan**

### CODE REVIEW

File: naming/dns_resolver.go [hunk 1]
```diff
@@ -192,28 +192,25 @@ type AddrMetadataGRPCLB struct {
...
-	update := make(map[Update]bool)
-	for _, u := range newAddrs {
-		update[*u] = true
-	}
-	for _, u := range w.curAddrs {
-		if _, ok := update[*u]; ok {
-			delete(update, *u)
-			continue
+func (w *dnsWatcher) compileUpdate(newAddrs map[string]*Update) []*Update {
+	res := make([]*Update, 0, len(newAddrs)+len(w.curAddrs))
```
Comment (menghanl) [inline naming/dns_resolver.go]: How about just `var res []*Update`?
Length of updates may be much shorter than `len(newAddrs)+len(w.curAddrs)`.

File: naming/dns_resolver.go [hunk 2]
```diff
@@ -231,15 +228,15 @@ func (w *dnsWatcher) lookupSRV() []*Update {
 				grpclog.Errorf("grpc: failed IP parsing due to %v.\n", err)
 				continue
 			}
-			newAddrs = append(newAddrs, &Update{Addr: a + ":" + strconv.Itoa(int(s.Port)),
-				Metadata: AddrMetadataGRPCLB{AddrType: GRPCLB, ServerName: s.Target}})
+			newAddrs[a+":"+strconv.Itoa(int(s.Port))] = &Update{Addr: a + ":" + strconv.Itoa(int(s.Port)),
```
Comment (menghanl) [inline naming/dns_resolver.go]: Add a local variable for `a+":"+strconv.Itoa(int(s.Port))`?

File: naming/dns_resolver.go [hunk 3]
```diff
@@ -251,7 +248,7 @@ func (w *dnsWatcher) lookupHost() []*Update {
 			grpclog.Errorf("grpc: failed IP parsing due to %v.\n", err)
 			continue
 		}
-		newAddrs = append(newAddrs, &Update{Addr: a + ":" + w.port})
+		newAddrs[a+":"+w.port] = &Update{Addr: a + ":" + w.port}
```
Comment (menghanl) [inline naming/dns_resolver.go]: Local variable for `a+":"+w.port`?

File: naming/dns_resolver.go (hunk 1 above)
Comment (lyuxuan) [inline naming/dns_resolver.go]: done

File: naming/dns_resolver.go [hunk 4]
```diff
@@ -192,28 +192,25 @@ type AddrMetadataGRPCLB struct {
...
-	for k := range update {
-		tmp := k
-		res = append(res, &tmp)
+	for a, u := range newAddrs {
+		if _, ok := w.curAddrs[a]; !ok {
+			res = append(res, u)
+		}
 	}
 	return res
+
```
Comment (lyuxuan) [inline naming/dns_resolver.go]: done

File: naming/dns_resolver.go (hunk 2 above)
Comment (lyuxuan) [inline naming/dns_resolver.go]: done

File: naming/dns_resolver.go (hunk 3 above)
Comment (lyuxuan) [inline naming/dns_resolver.go]: done

**Pull request merged**